### ISSUE TITLE
RNN layer: len is not well defined for a symbolic Tensor. Please call `x.shape` rather than `len(x)` for shape information.

### ISSUE THREAD

**Issue by ThibaultDECO** [keras-team/keras#19754]
I get the following error with keras:
`len is not well defined for a symbolic Tensor. Please call 'x.shape' rather than 'len(x)' for shape information.`

The issue seems to originate from [this line](https://github.com/keras-team/keras/blob/master/keras/src/layers/rnn/rnn.py#L431).

Here is [a gist](https://colab.research.google.com/gist/ThibaultDECO/a61f5719463e63b8c5f5ef3e5ee95ced/untitled1.ipynb) where the error is reproduced.

**Comment by ThibaultDECO** [comment]
Here are other places where a similar error message has been mentioned:
- [here](https://github.com/keras-team/keras/issues/15340)
- [here](https://github.com/tensorflow/tensorflow/issues/38908)
- [here](https://stackoverflow.com/questions/59682542/typeerror-len-is-not-well-defined-for-symbolic-tensors-activation-3-identity/59745136#59745136)

**Comment by SuryanarayanaY** [comment]
Hi @ThibaultDECO , Thanks for report. I have reproduced the issue with nightly and attached [gist](https://colab.sandbox.google.com/gist/SuryanarayanaY/674a0e8f91c0175fef6ce5abc67156d5/19754.ipynb) for reference.

**Comment by grasskin** [comment]
Hi @ThibaultDECO - thank you for flagging this. Would you be interested in contributing the fix? I believe just replacing `len(x)` with `len(ops.shape(x))` should work but you would have to confirm.

Otherwise we'll take this up.

**Comment by ThibaultDECO** [comment]
Hi @grasskin 
I unfortunately do not have time to look at the issue atm

**Comment by evz** [comment]
Just ran into something similar with the bidirectional layer type. Would [this line here](https://github.com/keras-team/keras/blob/458ece919c9b9157884dde36e0026205c20f3f99/keras/src/layers/rnn/bidirectional.py#L211) trigger this bug? The traceback I'm getting isn't super useful in this case so I'm just trying to figure out if I need to keep digging.

**Comment by ghsanti** [comment]
Did you disable traceback filtering to get the full traceback error? @evz

**Comment by evz** [comment]
Yeah so, I turned off the traceback filtering and it is, indeed, related. 

```
    File "/home/eric/code/coursera/.venv/lib/python3.10/site-packages/tensorflow/python/framework/tenso
r.py", line 633, in __len__
        raise TypeError(f"len is not well defined for a symbolic Tensor "

    TypeError: Exception encountered when calling Bidirectional.call().
     
    len is not well defined for a symbolic Tensor (states:0). Please call `x.shape` rather than `len(x)
` for shape information.
     
    Arguments received by Bidirectional.call():
      • sequences=tf.Tensor(shape=(1, None, 256), dtype=float32)
      • initial_state=tf.Tensor(shape=(1, 2048), dtype=float32)
      • mask=None
      • training=None
```
The backdrop here is that I'm working through the [Shakespeare RNN example](https://www.tensorflow.org/text/tutorials/text_generation) from the Tensorflow docs and toying around with adding a bidirectional layer to the model. I'm not even really sure it makes sense as a choice for that particular problem. I'm mostly just interested in figuring out how these things work.

**Comment by ghsanti** [comment]
I wonder whether x.shape is best or you'd still need `ops.shape(...)` and then `len(result)`.

I'm not sure about the details but i think `ops` namespace is to deal with symbolic tensors as it seems to be the case.

**Comment by ti55987** [comment]
I also encountered this issue with bi directional RNN.  Is the issue fixed?

**Comment by danielenricocahall** [comment]
In 2.14.0 this does not appear to be an issue anymore.

**Comment by dhantule** [comment]
Hi @ThibaultDECO, thanks for reporting this. 

I've tested your code with `Keras 3.13.2` , I'm no longer able to reproduce this error, everything seems to working fine, please refer this attached [gist](https://colab.sandbox.google.com/gist/dhantule/64dbb69c4a6a211415eef03f5c988717/rnn-layer-len-is-not-well-defined-for-a-symbolic-tensor-please-call-x-shape-rather-than-len-x-for-shape-information-19754.ipynb).
